### PR TITLE
[Merged by Bors] - Allow passing owned `HostHooks` and `JobQueues` to `Context`

### PR DESCRIPTION
**Pull request by jedel1043**
This allows `thread_local` contexts to have owned `HostHooks` and `JobQueues`.

It changes the following:

- Creates a new `MaybeShared` struct that can hold either a reference or an `Rc`.
- Changes the `job_queue` and `host_hooks` parameters of `Context` to use `MaybeShared`.

This PR also allows us to make `SimpleJobQueue` the default promise runner, which I think it's pretty cool :)

cc @lastmjs

### CODE REVIEW

**Review by Razican — APPROVED** [review]
Looks great!! Thanks!

I just had a small suggestion :)

File: boa_engine/src/context/mod.rs [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 mod hooks;
 pub mod intrinsics;
+mod maybe_shared;
 pub use hooks::{DefaultHooks, HostHooks};
+pub use maybe_shared::MaybeShared;
```
[the remark this answers is not in the document]
Comment (jedel1043) [reply boa_engine/src/context/mod.rs:7]: Uh so I added the attribute but I think it didn't change `rustdoc`'s output. Maybe it automatically inlines reexported structs?

[the remark this answers is not in the document]
Comment (Razican) [reply boa_engine/src/context/mod.rs]: It could be a default if their modules are not public. Let's resolve this :)

**Review by HalidOdat — APPROVED** [review]
Nice! Looks good to me! :)

**Pull request closed without merging**